apps/wpcom-block-editor/src/common/tracking.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * External dependencies
+ */
+import { use, select } from '@wordpress/data';
+import { registerPlugin } from '@wordpress/plugins';
+import { castArray } from 'lodash';
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import tracksRecordEvent from './tracking/track-record-event';
+import delegateEventTracking from './tracking/delegate-event-tracking';
+
+// Debugger.
+const debug = debugFactory( 'wpcom-block-editor:tracking' );
+
+// In case Tracks hasn't loaded.
+if ( typeof window !== 'undefined' ) {
+	window._tkq = window._tkq || [];
+}
+
+/**
+ * Looks up the block name based on its id.
+ *
+ * @param {string} blockId Blog identifier.
+ * @return {string|null} Blg name if it exists. Otherwise, `null`.
+ */
+const getTypeForBlockId = blockId => {
+	const block = select( 'core/block-editor' ).getBlock( blockId );
+	return block ? block.name : null;
+};
+
+/**
+ * A lot of actions accept either string (block id)
+ * or an array of multiple string when operating with multiple blocks selected.
+ * This is a convenience method that processes the first argument of the action (blocks)
+ * and calls your tracking for each of the blocks involved in the action.
+ *
+ * @param {string} eventName event name
+ * @return {function} track handler
+ */
+const getBlocksTracker = eventName => blockIds => {
+	// track separately for each block
+	castArray( blockIds ).forEach( blockId => {
+		tracksRecordEvent( eventName, { block_name: getTypeForBlockId( blockId ) } );
+	} );
+};
+
+/**
+ * Track block insertion.
+ *
+ * @param {object|array} blocks block instance object or an array of such objects
+ * @return {void}
+ */
+const trackBlockInsertion = blocks => {
+	castArray( blocks ).forEach( block => {
+		tracksRecordEvent( 'wpcom_block_picker_block_inserted', {
+			block_name: block.name,
+			blocks_replaced: false,
+		} );
+	} );
+};
```

Comment:
Block insertion is the event we're most interested in for the purpose of our `UpgradeNudge`, specifically to build funnels from block insertion to clicking the 'Upgrade' CTA (for which we already have an event, `jetpack_editor_block_upgrade_click`, see https://github.com/Automattic/jetpack/pull/12965).

Since we made the `UpgradeNudge` for Gutenberg blocks whose code lives in the Jetpack repo, and ship both with Jetpack and WP.com, our click event is prefixed with `jetpack_`. This sheds light on an interesting question: Even though we have activated the `UpgradeNudge` only on WP.com for now, we plan to eventually also activate it on Jetpack.

In that case, we'd obviously also like to build the same funnel, starting at block insertion! This means it'd be utterly useful to us if this event was also available inside Jetpack.

This can mean various things:
- Ideally, this code will eventually be available to both WP.com and Jetpack (npm etc, I'll elaborate later). But that's not even the most pressing question, since we can always move it around later.
- Naming is really important here! IMO, it'd kind of suck to have to build a funnel
  - from `wpcom_block_picker_block_inserted` to `jetpack_editor_block_upgrade_click` on WP.com, and a different one
  - from `jetpack_block_picker_block_inserted` to `jetpack_editor_block_upgrade_click` on Jetpack
(Remember that the 'bottom-of-the-funnel' event is the same (prefixed with `jetpack_`) in both cases!)

I think this would make analyzing our data considerably harder, and working with tracking much more confusing for anyone wanting to add events.

I think we have a number of options:
1. You folks stick with `wpcom_block_picker_block_inserted` (and we also call it that once we move it to Jetpack), we stick with `jetpack_editor_block_upgrade_click`. Clearly the most awkward, since funnels will be from `wpcom_` to `jetpack_` in both environments :man_facepalming: Super un-intuitive :no_entry_sign: 
2. You folks change yours to `jetpack_block_picker_block_inserted`, both on WP.com, and on Jetpack (and we stick with ours). Means consistency, but obviously awkward on WP.com. However, not more awkward than what we're doing with `jetpack_editor_block_upgrade_click`. For us, it was fairly easy to rationalize this decision, since after all, WP.com is Powered by JP, and more concretely, the same set of Gutenberg Blocks coming from Jetpack. Not sure you folks can get behind the same kind of reasoning :slightly_smiling_face: 
3. We make the prefixes dynamic: 
  - `wpcom_block_picker_block_inserted` and `wpcom_editor_block_upgrade_click` on WP.com
  - `jetpack_block_picker_block_inserted` to `jetpack_editor_block_upgrade_click` on Jetpack.

And to be clear, `wpcom_block_picker_block_inserted` just happens to be the one event at hand that we already know we're going to need, but I'm not suggesting special treatment for that event -- whatever option we choose, we need to apply it to all events we're adding to the block editor, or it'll be one giant mess with seemingly arbitrary prefixes.

So IMO, we should pick either option 2. or 3. I think it's easy to lean towards 3. (as I do, even though it means we have to change our event name post-launch :slightly_smiling_face: ), but we should really inform that decision by what is best for data analysis in the long run.

/cc @AnnaMag @sirreal @apeatling

Reply:
From PCYsg-4S2-p2 (recommended reading):

> Don’t programmatically create event names. They should be hardcoded so they are easily searchable in the code base.

I'd propose that events generated by the same code under the same circumstances are the same event. Whether they happen in WP Admin on a Simple, Atomic, or Jetpack site, they represent a user doing the same thing.

I'd propose that we use event properties to distinguish the type of site it's coming from. The tracking library may want to include a discriminating property with all events as part of its core functionality: `{ "site_type": "simple" | "atomic" | "jetpack" }`.

What should the `source` (the first section of the event name) be for these events? I'm not sure. According to PCYsg-4S2-p2:

> Prefix event names with the codebase or service generating the events […]

We don't have a single unified codebase. We'll need to decide on a tradeoff between knowing which codebase generates an event and folks working with funnels knowing that `(to be determined)_` tracks block insertion, while upgrade nudge clicks come from `jetpack_`.

Getting a new source whitelisted doesn't seem to be a problem.

Reply:
Thanks for your feedback on this - naming things is hard in the dev world. We already had long discussions internally and with the team behind Tracks. We initially wanted to have a new context whitelisted (`gutenberg_`) which was not approved.

`wpcom` was a suggestion to go with for reasons where this code runs today and it matched the precedent there is in the codebase, which is the [block search tracking](https://github.com/Automattic/wp-calypso/blob/master/apps/wpcom-block-editor/src/common/track-search.js#L62-L71).

The current plan is to go with `wpcom` as a source. If the tracking code ever makes it to other platforms, it will also use the exact same event name.

We are already sending through the blog id which can be used to determine the platform (wpcom/jetpack/atomic) as suggested by the Tracks team. I'm not sure if the tools you use for analysis allow you to simply filter platforms based on the blog id. If not, I think adding an additional property with the platform name is a good idea.

> I'd propose that we use event properties to distinguish the type of site it's coming from. The tracking library may want to include a discriminating property with all events as part of its core functionality: `{ "site_type": "simple" | "atomic" | "jetpack" }`.

This would be perfect I think.

Reply:
@retrofox has added `site_type` prop to our events with the value of `get_site_type( $current_site_id )`. At this moment "simple" is the only possible value.

Reply:
![image](https://user-images.githubusercontent.com/77539/61474585-58e68380-a95f-11e9-8d59-37a3e98602f0.png)

Reply:
Note that everything in the common folder will be served off of widgets.wp.com and is available for all site types (Simple, Jetpack, Atomic) in wp-admin + WordPress.com/block-editor

Reply:
Verified that I see tracking events in Simple and Jetpack with the patch applied. Caches are invalidated daily, but this will be served to all site types. See:

https://github.com/Automattic/jetpack/blob/bb9aaeb930ada902d698060f5f6e45b689df3d55/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php#L259

Reply:
In the long run I think it may make more sense to

1. Pull out analytics npm package as part of the Calypso monorepo
2. Remove direct Block-Editor tracking calls in Jetpack
3. Setup an analytics custom store that can listen for redux events. Components can fire a redux event, and redux middleware can make the actual tracking call. 
3. In terms of tracks namespace, let's pick something consistent, it doesn't really matter where the tracking code lives. +1 on Jon's suggestion to provide a site type source + blog id and for searchable names.

cc @stambizzle in case you had more suggestions on names / best practices.

Reply:
Ignore most of this last comment :D I wrote this when I wasn't far down enough in the PR 👍